java.lang.IllegalStateException: cannot consume EOF

The following code produces an exception when parsing:

```sql
IF EXISTS (SELECT 1 FROM dbo.sysobjects WHERE id = object_id(N'pk_t') AND OBJECTPROPERTY(id, N'IsPrimaryKey') = 1)
ALTER TABLE [t] DROP CONSTRAINT [pk_t]
GO

--IF EXISTS (SELECT 1 FROM dbo.sysobjects WHERE id = object_id(N'[pk_t2]')  
--DROP TABLE [t2]
```

Here's the error stack:

```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:152)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:220)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:187)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.reflect.InvocationTargetException
	at jdk.internal.reflect.GeneratedMethodAccessor2.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 11 more
Caused by: java.lang.IllegalStateException: cannot consume EOF
	at org.antlr.v4.runtime.CodePointCharStream.consume(CodePointCharStream.java:96)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.consume(LexerATNSimulator.java:742)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.evaluatePredicate(LexerATNSimulator.java:592)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.getEpsilonTarget(LexerATNSimulator.java:507)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:453)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:455)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:455)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:455)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:455)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:455)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:455)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:455)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:435)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:455)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:435)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:455)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:455)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.closure(LexerATNSimulator.java:455)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.getReachableConfigSet(LexerATNSimulator.java:342)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.computeTargetState(LexerATNSimulator.java:277)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.execATN(LexerATNSimulator.java:204)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.matchATN(LexerATNSimulator.java:156)
	at org.antlr.v4.runtime.atn.LexerATNSimulator.match(LexerATNSimulator.java:111)
	at org.antlr.v4.runtime.Lexer.nextToken(Lexer.java:141)
	at org.antlr.v4.runtime.BufferedTokenStream.fetch(BufferedTokenStream.java:169)
	at org.antlr.v4.runtime.BufferedTokenStream.fill(BufferedTokenStream.java:485)
	at ch.islandsql.grammar.util.TokenStreamUtil.hideOutOfScopeTokens(TokenStreamUtil.java:42)
	at ch.islandsql.grammar.IslandSqlDocument.<init>(IslandSqlDocument.java:49)
	at ch.islandsql.grammar.IslandSqlDocument.parse(IslandSqlDocument.java:67)
	at ch.islandsql.ls.IslandSqlTextDocumentService.didChange(IslandSqlTextDocumentService.java:106)
	... 15 more
```

Culprit is probably `at ch.islandsql.grammar.util.TokenStreamUtil.hideOutOfScopeTokens(TokenStreamUtil.java:42)`

Here's the screenshot from VS Code:

<img width="903" alt="image" src="https://user-images.githubusercontent.com/3883171/223099836-c25d8e3d-3806-492c-976e-b50a62f048f8.png">

A related stack is produce when running a search in the VS Code extension.

Here's an example:

```
SEVERE: Internal error: java.lang.NullPointerException: Cannot invoke "ch.islandsql.grammar.IslandSqlDocument.getAllContentsOfType(java.lang.Class)" because "doc" is null
java.util.concurrent.CompletionException: java.lang.NullPointerException: Cannot invoke "ch.islandsql.grammar.IslandSqlDocument.getAllContentsOfType(java.lang.Class)" because "doc" is null
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1770)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: java.lang.NullPointerException: Cannot invoke "ch.islandsql.grammar.IslandSqlDocument.getAllContentsOfType(java.lang.Class)" because "doc" is null
	at ch.islandsql.ls.IslandSqlTextDocumentService.lambda$documentSymbol$1(IslandSqlTextDocumentService.java:231)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	... 6 more

[Error - 12:38:28] Request textDocument/documentSymbol failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: java.lang.NullPointerException: Cannot invoke "ch.islandsql.grammar.IslandSqlDocument.getAllContentsOfType(java.lang.Class)" because "doc" is null
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1770)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: java.lang.NullPointerException: Cannot invoke "ch.islandsql.grammar.IslandSqlDocument.getAllContentsOfType(java.lang.Class)" because "doc" is null
	at ch.islandsql.ls.IslandSqlTextDocumentService.lambda$documentSymbol$1(IslandSqlTextDocumentService.java:231)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	... 6 more
```

--- a/pom.xml
+++ b/pom.xml
@@ -5,7 +5,7 @@
     <!-- The Basics -->
     <groupId>ch.islandsql</groupId>
     <artifactId>islandsql</artifactId>
-    <version>0.5.0-SNAPSHOT</version>
+    <version>0.4.1-SNAPSHOT</version>
     <packaging>jar</packaging>
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>

--- a/src/main/antlr4/ch/islandsql/grammar/IslandSqlLexer.g4
+++ b/src/main/antlr4/ch/islandsql/grammar/IslandSqlLexer.g4
@@ -214,7 +214,7 @@ MERGE:
 SELECT:
       (
           // TODO: remove alternative with https://github.com/IslandSQL/IslandSQL/issues/29
-          ('(' COMMENT_OR_WS*)+ ('select'|'with') .*? (')' COMMENT_OR_WS*)+ {isLoop()}?
+          ('(' COMMENT_OR_WS*)+ ('select'|'with') .*? (')' COMMENT_OR_WS*)+ {isText("loop")}?
       )
     | (
           (

--- a/src/main/antlr4/ch/islandsql/grammar/IslandSqlLexer.g4
+++ b/src/main/antlr4/ch/islandsql/grammar/IslandSqlLexer.g4
@@ -45,12 +45,12 @@ fragment INT: [0-9]+;
 /*----------------------------------------------------------------------------*/
 
 REMARK_COMMAND:
-    {isBeginOfCommand()}? 'rem' ('a' ('r' 'k'?)?)?
+    'rem' {isBeginOfCommand("rem")}? ('a' ('r' 'k'?)?)?
         (WS SQLPLUS_TEXT*)? SQLPLUS_END -> channel(HIDDEN)
 ;
 
 PROMPT_COMMAND:
-    {isBeginOfCommand()}? 'pro' ('m' ('p' 't'?)?)?
+    'pro' {isBeginOfCommand("pro")}? ('m' ('p' 't'?)?)?
        (WS SQLPLUS_TEXT*)? SQLPLUS_END -> channel(HIDDEN)
 ;
 
@@ -188,45 +188,50 @@ QUOTED_ID: '"' .*? '"' ('"' .*? '"')*;
 ID: [\p{Alpha}] [_$#0-9\p{Alpha}]*;
 
 /*----------------------------------------------------------------------------*/
+// Cursor for loop
+// TODO: remove with https://github.com/IslandSQL/IslandSQL/issues/29
+/*----------------------------------------------------------------------------*/
+
+CURSOR_FOR_LOOP_START:
+    'for' COMMENT_OR_WS+ ~[\t\r\n ]+ COMMENT_OR_WS+ 'in' COMMENT_OR_WS* {isText("(")}?
+    -> channel(HIDDEN), pushMode(CURSOR_FOR_LOOP)
+;
+
+/*----------------------------------------------------------------------------*/
 // Islands of interest as single tokens
 /*----------------------------------------------------------------------------*/
 
 CALL:
-    'call' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
+    'call' {isBeginOfStatement("call")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END
 ;
 
 DELETE:
-    'delete' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
+    'delete' {isBeginOfStatement("delete")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END
 ;
 
 EXPLAIN_PLAN:
-    'explain' COMMENT_OR_WS+ 'plan' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
+    'explain' {isBeginOfStatement("explain")}? COMMENT_OR_WS+ 'plan' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
 ;
 
 INSERT:
-    'insert' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
+    'insert' {isBeginOfStatement("insert")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END
 ;
 
 MERGE:
-    'merge' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
+    'merge' {isBeginOfStatement("merge")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END
 ;
 
 SELECT:
-      (
-          // TODO: remove alternative with https://github.com/IslandSQL/IslandSQL/issues/29
-          ('(' COMMENT_OR_WS*)+ ('select'|'with') .*? (')' COMMENT_OR_WS*)+ {isText("loop")}?
-      )
-    | (
-          (
-                ('with' COMMENT_OR_WS+ ('function'|'procedure') SQL_TEXT+? PLSQL_DECLARATION_END)
-              | ('with' COMMENT_OR_WS+ SQL_TEXT+? SQL_END)
-              | (('(' COMMENT_OR_WS*)* 'select' COMMENT_OR_WS+ SQL_TEXT+? SQL_END)
-          )
-      )
+    (
+        ('with' {isBeginOfStatement("with")}? COMMENT_OR_WS+ ('function'|'procedure') SQL_TEXT+? PLSQL_DECLARATION_END)
+      | ('with' {isBeginOfStatement("with")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END)
+      | ('select' {isBeginOfStatement("select")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END)
+      | ('(' {isBeginOfStatement("(")}? COMMENT_OR_WS? ('(' COMMENT_OR_WS*)* 'select' COMMENT_OR_WS+ SQL_TEXT+? SQL_END)
+    )
 ;
 
 UPDATE:
-    'update' COMMENT_OR_WS+ SQL_TEXT+? 'set' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
+    'update' {isBeginOfStatement("update")}? COMMENT_OR_WS+ SQL_TEXT+? 'set' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
 ;
 
 /*----------------------------------------------------------------------------*/
@@ -234,3 +239,26 @@ UPDATE:
 /*----------------------------------------------------------------------------*/
 
 ANY_OTHER: .;
+
+/*----------------------------------------------------------------------------*/
+// Cursor for loop mode to identify select statement
+// TODO: remove with https://github.com/IslandSQL/IslandSQL/issues/29
+/*----------------------------------------------------------------------------*/
+
+mode CURSOR_FOR_LOOP;
+
+fragment CFL_SINGLE_NL: '\r'? '\n';
+fragment CFL_COMMENT_OR_WS: CFL_ML_COMMENT|CFL_SL_COMMENT|CFL_WS;
+CFL_ML_COMMENT: '/*' .*? '*/' -> channel(HIDDEN), type(ML_COMMENT);
+CFL_SL_COMMENT: '--' .*? (EOF|CFL_SINGLE_NL) -> channel(HIDDEN), type(SL_COMMENT);
+CFL_WS: [ \t\r\n]+ -> channel(HIDDEN), type(WS);
+CFL_ANY_OTHER: . -> channel(HIDDEN), type(ANY_OTHER);
+
+CFL_SELECT:
+    ('(' CFL_COMMENT_OR_WS*)+
+    ('select'|'with') .*? (')' CFL_COMMENT_OR_WS*)+ {isText("loop")}? -> type(SELECT)
+;
+
+CFL_END_OF_SELECT:
+    'loop' -> channel(HIDDEN), popMode
+;

--- a/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
+++ b/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
@@ -84,7 +84,7 @@ REVOKE:
 /*----------------------------------------------------------------------------*/
 
 CURSOR_FOR_LOOP_START:
-    'for' COMMENT_OR_WS+ SQL_TEXT+? 'in' COMMENT_OR_WS*
+    'for' COMMENT_OR_WS+ SQL_TEXT+? COMMENT_OR_WS+ 'in' COMMENT_OR_WS*
     -> channel(HIDDEN), pushMode(CURSOR_FOR_LOOP)
 ;
 

--- a/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
+++ b/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
@@ -47,12 +47,12 @@ ML_COMMENT: '/*' .*? '*/' -> channel(HIDDEN);
 SL_COMMENT: '--' (~[\r\n])* (EOF|SINGLE_NL) -> channel(HIDDEN);
 
 REMARK_COMMAND:
-    {isBeginOfCommand()}? 'rem' ('a' ('r' 'k'?)?)?
+    'rem' {isBeginOfCommand("rem")}? ('a' ('r' 'k'?)?)?
         (WS SQLPLUS_TEXT*)? SQLPLUS_END -> channel(HIDDEN)
 ;
 
 PROMPT_COMMAND:
-    {isBeginOfCommand()}? 'pro' ('m' ('p' 't'?)?)?
+    'pro' {isBeginOfCommand("pro")}?  ('m' ('p' 't'?)?)?
        (WS SQLPLUS_TEXT*)? SQLPLUS_END -> channel(HIDDEN)
 ;
 
@@ -71,11 +71,11 @@ STRING:
 CONDITIONAL_COMPILATION_DIRECTIVE: '$if' .*? '$end' -> channel(HIDDEN);
 
 GRANT:
-    {isBeginOfStatement()}? 'grant' COMMENT_OR_WS+ SQL_TEXT+? SQL_END -> channel(HIDDEN)
+    'grant' {isBeginOfStatement("grant")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END -> channel(HIDDEN)
 ;
 
 REVOKE:
-    {isBeginOfStatement()}? 'revoke' COMMENT_OR_WS+ SQL_TEXT+? SQL_END -> channel(HIDDEN)
+    'revoke' {isBeginOfStatement("revoke")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END -> channel(HIDDEN)
 ;
 
 /*----------------------------------------------------------------------------*/
@@ -93,40 +93,40 @@ CURSOR_FOR_LOOP_START:
 /*----------------------------------------------------------------------------*/
 
 CALL:
-    {isBeginOfStatement()}? 'call' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
+    'call' {isBeginOfStatement("call")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END
 ;
 
 DELETE:
-    {isBeginOfStatement()}? 'delete' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
+    'delete' {isBeginOfStatement("delete")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END
 ;
 
 EXPLAIN_PLAN:
-    {isBeginOfStatement()}? 'explain' COMMENT_OR_WS+ 'plan' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
+    'explain' {isBeginOfStatement("explain")}? COMMENT_OR_WS+ 'plan' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
 ;
 
 INSERT:
-    {isBeginOfStatement()}? 'insert' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
+    'insert' {isBeginOfStatement("insert")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END
 ;
 
 LOCK_TABLE:
-    {isBeginOfStatement()}? 'lock' COMMENT_OR_WS+ 'table' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
+    'lock' {isBeginOfStatement("lock")}? COMMENT_OR_WS+ 'table' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
 ;
 
 MERGE:
-    {isBeginOfStatement()}? 'merge' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
+    'merge' {isBeginOfStatement("merge")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END
 ;
 
 SELECT:
-    {isBeginOfStatement()}?
     (
-        ('with' COMMENT_OR_WS+ ('function'|'procedure') SQL_TEXT+? PLSQL_DECLARATION_END)
-      | ('with' COMMENT_OR_WS+ SQL_TEXT+? SQL_END)
-      | (('(' COMMENT_OR_WS*)* 'select' COMMENT_OR_WS+ SQL_TEXT+? SQL_END)
+        ('with' {isBeginOfStatement("with")}? COMMENT_OR_WS+ ('function'|'procedure') SQL_TEXT+? PLSQL_DECLARATION_END)
+      | ('with' {isBeginOfStatement("with")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END)
+      | ('select' {isBeginOfStatement("select")}? COMMENT_OR_WS+ SQL_TEXT+? SQL_END)
+      | ('(' {isBeginOfStatement("(")}? COMMENT_OR_WS? ('(' COMMENT_OR_WS*)* 'select' COMMENT_OR_WS+ SQL_TEXT+? SQL_END)
     )
 ;
 
 UPDATE:
-    {isBeginOfStatement()}? 'update' COMMENT_OR_WS+ SQL_TEXT+? 'set' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
+    'update' {isBeginOfStatement("update")}? COMMENT_OR_WS+ SQL_TEXT+? 'set' COMMENT_OR_WS+ SQL_TEXT+? SQL_END
 ;
 
 /*----------------------------------------------------------------------------*/

--- a/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
+++ b/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
@@ -44,7 +44,7 @@ fragment SQLPLUS_END: EOF|SINGLE_NL;
 /*----------------------------------------------------------------------------*/
 
 ML_COMMENT: '/*' .*? '*/' -> channel(HIDDEN);
-SL_COMMENT: '--' .*? (EOF|SINGLE_NL) -> channel(HIDDEN);
+SL_COMMENT: '--' (~[\r\n])* (EOF|SINGLE_NL) -> channel(HIDDEN);
 
 REMARK_COMMAND:
     {isBeginOfCommand()}? 'rem' ('a' ('r' 'k'?)?)?

--- a/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
+++ b/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
@@ -157,7 +157,7 @@ CFL_ANY_OTHER: . -> channel(HIDDEN), type(ANY_OTHER);
 
 CFL_SELECT:
     ('(' CFL_COMMENT_OR_WS*)+
-    ('select'|'with') .*? (')' CFL_COMMENT_OR_WS*)+ {isLoop()}? -> type(SELECT)
+    ('select'|'with') .*? (')' CFL_COMMENT_OR_WS*)+ {isText("loop")}? -> type(SELECT)
 ;
 
 CFL_END_OF_SELECT:

--- a/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
+++ b/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
@@ -84,7 +84,7 @@ REVOKE:
 /*----------------------------------------------------------------------------*/
 
 CURSOR_FOR_LOOP_START:
-    'for' COMMENT_OR_WS+ SQL_TEXT+? COMMENT_OR_WS+ 'in' COMMENT_OR_WS*
+    'for' COMMENT_OR_WS+ SQL_TEXT+? COMMENT_OR_WS+ 'in' COMMENT_OR_WS* {isText("(")}?
     -> channel(HIDDEN), pushMode(CURSOR_FOR_LOOP)
 ;
 

--- a/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
+++ b/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
@@ -153,7 +153,7 @@ fragment CFL_COMMENT_OR_WS: CFL_ML_COMMENT|CFL_SL_COMMENT|CFL_WS;
 CFL_ML_COMMENT: '/*' .*? '*/' -> channel(HIDDEN), type(ML_COMMENT);
 CFL_SL_COMMENT: '--' .*? (EOF|CFL_SINGLE_NL) -> channel(HIDDEN), type(SL_COMMENT);
 CFL_WS: [ \t\r\n]+ -> channel(HIDDEN), type(WS);
-CFL_ANY_OTHER: '.' -> channel(HIDDEN), type(ANY_OTHER);
+CFL_ANY_OTHER: . -> channel(HIDDEN), type(ANY_OTHER);
 
 CFL_SELECT:
     ('(' CFL_COMMENT_OR_WS*)+

--- a/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
+++ b/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4
@@ -84,7 +84,7 @@ REVOKE:
 /*----------------------------------------------------------------------------*/
 
 CURSOR_FOR_LOOP_START:
-    'for' COMMENT_OR_WS+ SQL_TEXT+? COMMENT_OR_WS+ 'in' COMMENT_OR_WS* {isText("(")}?
+    'for' COMMENT_OR_WS+ ~[\t\r\n ]+ COMMENT_OR_WS+ 'in' COMMENT_OR_WS* {isText("(")}?
     -> channel(HIDDEN), pushMode(CURSOR_FOR_LOOP)
 ;
 

--- a/src/main/java/ch/islandsql/grammar/IslandSqlDocument.java
+++ b/src/main/java/ch/islandsql/grammar/IslandSqlDocument.java
@@ -45,12 +45,12 @@ public class IslandSqlDocument {
     private IslandSqlDocument(String sql) {
         CodePointCharStream charStream = CharStreams.fromString(sql);
         IslandSqlLexer lexer = new IslandSqlLexer(charStream);
-        this.tokenStream = new CommonTokenStream(lexer);
-        TokenStreamUtil.hideOutOfScopeTokens(tokenStream);
-        IslandSqlParser parser = new IslandSqlParser(tokenStream);
         SyntaxErrorListener errorListener = new SyntaxErrorListener();
         lexer.removeErrorListeners();
         lexer.addErrorListener(errorListener);
+        this.tokenStream = new CommonTokenStream(lexer);
+        TokenStreamUtil.hideOutOfScopeTokens(tokenStream, errorListener);
+        IslandSqlParser parser = new IslandSqlParser(tokenStream);
         parser.removeErrorListeners();
         parser.addErrorListener(errorListener);
         this.file = parser.file();

--- a/src/main/java/ch/islandsql/grammar/IslandSqlLexerBase.java
+++ b/src/main/java/ch/islandsql/grammar/IslandSqlLexerBase.java
@@ -87,16 +87,17 @@ public abstract class IslandSqlLexerBase extends Lexer {
     }
 
     /**
-     * Determines if the current position is valid for a command.
+     * Determines if the position beforeString is valid for a command.
      * A command must start on a new line. Between the new line and
      * the start of command an unbounded number of whitespace is
      * allowed. In other words a command can start in any column.
      * A command can start at begin-of-file.
      *
+     * @param beforeString String used to determine start of command.
      * @return Returns true if the current position is valid for a command.
      */
-    public boolean isBeginOfCommand() {
-        int i = _input.index() - 1;
+    public boolean isBeginOfCommand(String beforeString) {
+        int i = _input.index() - beforeString.length() - 1;
         while (isCharOneOf(" \t\r\n", i)) {
             if (i < 0 || isCharOneOf("\r\n", i)) {
                 return true;
@@ -107,15 +108,16 @@ public abstract class IslandSqlLexerBase extends Lexer {
     }
 
     /**
-     * Determines if the current position is valid for a SQL statement.
+     * Determines if the position beforeString is valid for a SQL statement.
      * A SQL statement starts after a whitespace or a semicolon.
      * In other words multiple SQL statements on a single line are allowed.
      * A SQL statement can start at begin-of-file.
      *
+     * @param beforeString String used to determine start of command.
      * @return Returns true if the current position is valid for a SQL statement.
      */
-    public boolean isBeginOfStatement() {
-        return isCharOneOf(" \t\r\n;", _input.index()-1);
+    public boolean isBeginOfStatement(String beforeString) {
+        return isCharOneOf(" \t\r\n;", _input.index() - beforeString.length() -1);
     }
 
     /**

--- a/src/main/java/ch/islandsql/grammar/IslandSqlLexerBase.java
+++ b/src/main/java/ch/islandsql/grammar/IslandSqlLexerBase.java
@@ -37,12 +37,13 @@ public abstract class IslandSqlLexerBase extends Lexer {
     }
 
     /**
-     * Determines if "loop" is the next word in the character stream.
+     * Determines if text is the next character sequencein the character stream.
      *
-     * @return Returns true if "loop" is the next word in the character stream.
+     * @return Returns true if text is the next character sequence in the character stream.
      */
-    public boolean isLoop() {
-        return "loop".equalsIgnoreCase(getInputStream().getText(Interval.of(_input.index(), _input.index() + 3)));
+    public boolean isText(String text) {
+        return text.equalsIgnoreCase(getInputStream()
+                .getText(Interval.of(_input.index(), _input.index() + text.length() - 1)));
     }
 
     /**

--- a/src/main/java/ch/islandsql/grammar/util/TokenStreamUtil.java
+++ b/src/main/java/ch/islandsql/grammar/util/TokenStreamUtil.java
@@ -59,14 +59,24 @@ public class TokenStreamUtil {
             scopeLexer.addErrorListener(errorListener);
         }
         CommonTokenStream scopeStream = new CommonTokenStream(scopeLexer);
-        scopeStream.fill();
+        try {
+            scopeStream.fill();
+        } catch (IllegalStateException e) {
+            // best effort
+        }
         List<Token> scopeTokens = new ArrayList<>(scopeStream.getTokens());
         int scopeIndex = 0;
         Token scopeToken = scopeTokens.get(scopeIndex);
+        tokenLoop:
         for (CommonToken token : tokens) {
             while (scopeToken.getType() != Token.EOF && scopeToken.getStopIndex() < token.getStartIndex()) {
                 scopeIndex++;
-                scopeToken = scopeTokens.get(scopeIndex);
+                try {
+                    scopeToken = scopeTokens.get(scopeIndex);
+                } catch (IndexOutOfBoundsException e) {
+                    // best effort
+                    break tokenLoop;
+                }
             }
             if (token.getChannel() != Token.HIDDEN_CHANNEL &&
                     (scopeToken.getChannel() == Token.HIDDEN_CHANNEL || scopeToken.getType() == Token.EOF) ) {

--- a/src/main/java/ch/islandsql/grammar/util/TokenStreamUtil.java
+++ b/src/main/java/ch/islandsql/grammar/util/TokenStreamUtil.java
@@ -39,7 +39,18 @@ public class TokenStreamUtil {
      * @param tokenStream The tokensStream produced by islandSqlLexer to process.
      */
     static public void hideOutOfScopeTokens(CommonTokenStream tokenStream, SyntaxErrorListener errorListener) {
-        tokenStream.fill();
+        try {
+            tokenStream.fill();
+        } catch (IllegalStateException e) {
+            // Workaround for issue 44 ("cannot consume EOF"). ATM no idea how to solve it.
+            // I see two options.
+            //   a) abort process and return; then syntax errors are reported
+            //   b) just ignore the exception and suppress the syntax errors
+            // Both options have pros and cons. In any case the number of tokens are incomplete.
+            // Suppressing the syntax errors is not optimal, but it's probably better than
+            // reporting errors for SQL scripts of out-of-scope dialects.
+            // TODO: add "return;" with https://github.com/IslandSQL/IslandSQL/issues/21
+        }
         List<CommonToken> tokens = tokenStream.getTokens().stream().map(t -> (CommonToken)t).collect(Collectors.toList());
         CodePointCharStream charStream = CharStreams.fromString(tokenStream.getText());
         IslandSqlScopeLexer scopeLexer = new IslandSqlScopeLexer(charStream);

--- a/src/main/java/ch/islandsql/grammar/util/TokenStreamUtil.java
+++ b/src/main/java/ch/islandsql/grammar/util/TokenStreamUtil.java
@@ -43,13 +43,25 @@ public class TokenStreamUtil {
             tokenStream.fill();
         } catch (IllegalStateException e) {
             // Workaround for issue 44 ("cannot consume EOF"). ATM no idea how to solve it.
-            // I see two options.
+            // I see the following options:
             //   a) abort process and return; then syntax errors are reported
             //   b) just ignore the exception and suppress the syntax errors
-            // Both options have pros and cons. In any case the number of tokens are incomplete.
-            // Suppressing the syntax errors is not optimal, but it's probably better than
-            // reporting errors for SQL scripts of out-of-scope dialects.
-            // TODO: add "return;" with https://github.com/IslandSQL/IslandSQL/issues/21
+            //   c) ignore the exception and report the issue as syntax error
+            // I chose c). However, the best would be to identify why this happens and to
+            // fix the lexer. Reporting the error is the best option ATM and should help
+            // to identify the root cause in the lexer.
+            if (errorListener != null) {
+                Token offendingToken = null;
+                int size = tokenStream.size();
+                int line = 0;
+                int charPositionInLine = 0;
+                if (size > 0) {
+                    offendingToken = tokenStream.get(size-1);
+                    line = offendingToken.getLine();
+                    charPositionInLine = offendingToken.getCharPositionInLine();
+                }
+                errorListener.syntaxError(null, offendingToken, line, charPositionInLine, e.getMessage() + " (IslandSqlLexer)", null);
+            }
         }
         List<CommonToken> tokens = tokenStream.getTokens().stream().map(t -> (CommonToken)t).collect(Collectors.toList());
         CodePointCharStream charStream = CharStreams.fromString(tokenStream.getText());
@@ -62,7 +74,19 @@ public class TokenStreamUtil {
         try {
             scopeStream.fill();
         } catch (IllegalStateException e) {
-            // best effort
+            // best effort, similar issue as for tokenStream, probably related.
+            if (errorListener != null) {
+                Token offendingToken = null;
+                int size = scopeStream.size();
+                int line = 0;
+                int charPositionInLine = 0;
+                if (size > 0) {
+                    offendingToken = scopeStream.get(size-1);
+                    line = offendingToken.getLine();
+                    charPositionInLine = offendingToken.getCharPositionInLine();
+                }
+                errorListener.syntaxError(null, offendingToken, line, charPositionInLine, e.getMessage() + " (IslandSqlScopeLexer)", null);
+            }
         }
         List<Token> scopeTokens = new ArrayList<>(scopeStream.getTokens());
         int scopeIndex = 0;
@@ -74,7 +98,9 @@ public class TokenStreamUtil {
                 try {
                     scopeToken = scopeTokens.get(scopeIndex);
                 } catch (IndexOutOfBoundsException e) {
-                    // best effort
+                    // best effort, subsequent error of previous IllegalStateException
+                    // no need to report this error, just stop the processing,
+                    // the parser will probably produce further subsequent errors.
                     break tokenLoop;
                 }
             }

--- a/src/main/java/ch/islandsql/grammar/util/TokenStreamUtil.java
+++ b/src/main/java/ch/islandsql/grammar/util/TokenStreamUtil.java
@@ -42,14 +42,8 @@ public class TokenStreamUtil {
         try {
             tokenStream.fill();
         } catch (IllegalStateException e) {
-            // Workaround for issue 44 ("cannot consume EOF"). ATM no idea how to solve it.
-            // I see the following options:
-            //   a) abort process and return; then syntax errors are reported
-            //   b) just ignore the exception and suppress the syntax errors
-            //   c) ignore the exception and report the issue as syntax error
-            // I chose c). However, the best would be to identify why this happens and to
-            // fix the lexer. Reporting the error is the best option ATM and should help
-            // to identify the root cause in the lexer.
+            // Fail-safe for issues like #44 ("cannot consume EOF").
+            // Syntax error is reported. This helps to identify the root cause in the lexer.
             if (errorListener != null) {
                 Token offendingToken = null;
                 int size = tokenStream.size();
@@ -74,7 +68,8 @@ public class TokenStreamUtil {
         try {
             scopeStream.fill();
         } catch (IllegalStateException e) {
-            // best effort, similar issue as for tokenStream, probably related.
+            // Fail save for issues in the lexer.
+            // Syntax error is reported. This helps to identify the root cause in the lexer.
             if (errorListener != null) {
                 Token offendingToken = null;
                 int size = scopeStream.size();
@@ -101,6 +96,7 @@ public class TokenStreamUtil {
                     // best effort, subsequent error of previous IllegalStateException
                     // no need to report this error, just stop the processing,
                     // the parser will probably produce further subsequent errors.
+                    // the root cause in the lexer needs to be identified and fixed.
                     break tokenLoop;
                 }
             }

--- a/src/main/java/ch/islandsql/grammar/util/TokenStreamUtil.java
+++ b/src/main/java/ch/islandsql/grammar/util/TokenStreamUtil.java
@@ -38,11 +38,15 @@ public class TokenStreamUtil {
      *
      * @param tokenStream The tokensStream produced by islandSqlLexer to process.
      */
-    static public void hideOutOfScopeTokens(CommonTokenStream tokenStream) {
+    static public void hideOutOfScopeTokens(CommonTokenStream tokenStream, SyntaxErrorListener errorListener) {
         tokenStream.fill();
         List<CommonToken> tokens = tokenStream.getTokens().stream().map(t -> (CommonToken)t).collect(Collectors.toList());
         CodePointCharStream charStream = CharStreams.fromString(tokenStream.getText());
         IslandSqlScopeLexer scopeLexer = new IslandSqlScopeLexer(charStream);
+        if (errorListener != null) {
+            scopeLexer.removeErrorListeners();
+            scopeLexer.addErrorListener(errorListener);
+        }
         CommonTokenStream scopeStream = new CommonTokenStream(scopeLexer);
         scopeStream.fill();
         List<Token> scopeTokens = new ArrayList<>(scopeStream.getTokens());

